Make autogen functions correct for multiple outputs and views

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31990 Make autogen functions correct for multiple outputs and views**
* #32044 fix view listing in autograd codegen

This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output. Introduce a whitelist of functions for which we allow this.

Differential Revision: [D19409649](https://our.internmc.facebook.com/intern/diff/D19409649)

albanD committed:
Make autogen functions correct for multiple outputs and views

[ghstack-poisoned]

ezyang commented:
Any more description than just the title?

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 32d61d71:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=32d61d71d56f07b7a733b5b057d34df7da15ff5b).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4184796) pytorch_linux_xenial_py3_6_gcc5_4_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=82667627) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=82667627))
<details>
<summary>
<code>Jan 09 17:24:24 RuntimeError: test_autograd failed!</code>
</summary>

```
Jan 09 17:24:24 Ran 1032 tests in 211.084s 
Jan 09 17:24:24  
Jan 09 17:24:24 FAILED (failures=2, skipped=12, expected failures=1) 
Jan 09 17:24:24  
Jan 09 17:24:24 Generating XML reports... 
Jan 09 17:24:24 Traceback (most recent call last): 
Jan 09 17:24:24   File "test/run_test.py", line 456, in <module> 
Jan 09 17:24:24     main() 
Jan 09 17:24:24   File "test/run_test.py", line 449, in main 
Jan 09 17:24:24     raise RuntimeError(message) 
Jan 09 17:24:24 RuntimeError: test_autograd failed! 
Jan 09 17:24:24 =================== sccache compilation log =================== 
Jan 09 17:24:24 + cleanup 
Jan 09 17:24:24 + retcode=1 
Jan 09 17:24:24 + set +x 
Jan 09 17:24:25 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 09 17:24:25 Compile requests                  7 
Jan 09 17:24:25 Compile requests executed         6 
Jan 09 17:24:25 Cache hits                        0 
Jan 09 17:24:25 Cache misses                      6 
Jan 09 17:24:25 Cache timeouts                    0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This makes the autogen functions properly set the allow_rebase_history flag to ensure that they don't silently give incorrect gradients.

Will provide a diff from the codegen as soon as I get it.

[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/compare/0a53609...albanD:codegen_diff).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/compare/0a53609...albanD:codegen_diff).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/compare/0a53609...albanD:codegen_diff).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/compare/0a53609...albanD:codegen_diff).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/compare/0a53609...albanD:codegen_diff).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/compare/0a53609...albanD:codegen_diff).


[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -145,6 +145,10 @@ namespace impl {
 
+  /// Mark a view such that its history cannot be rebased.
+  /// This is only valid to call on views.
+  TORCH_API void disable_rebase_history(Variable&);
```
Is this something that morally is only called at view construction time, or can a Variable have its rebase history disabled at some later point in time? When would I call this function?

ezyang replied:
If it is only called at construction time, can it be moved into the constructor directly?

albanD replied:
After offline discussion. No. So this is moved to the constructor. And the other functions are cleaned-up to reduce the number of boolean arguments.

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -76,6 +77,7 @@ namespace impl {
     if (self.is_view()) {
       // NB: is_view() ==> get_autograd_meta()
       auto diff_view_meta = static_cast<DifferentiableViewMeta*>(get_autograd_meta(self));
+      AT_ASSERT(diff_view_meta->allow_rebase_history);
```
nit: `AT_ASSERT` is deprecated, use `TORCH_INTERNAL_ASSERT` if this case can never be triggered by user code.

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -91,6 +93,13 @@ namespace impl {
   }
 
+  void disable_rebase_history(Variable& self) {
+    TORCH_CHECK(self.is_view(), "disable_rebase_history can only be called on views.");
```
Is this a CHECK or an ASSERT? This doesn't seem like public API surface to me

albanD replied:
That would be and assert. I'll change it to `TORCH_INTERNAL_ASSERT` as well.

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -41,9 +41,18 @@ namespace torch { namespace autograd {
+      auto diff_view_meta = static_cast<DifferentiableViewMeta*>(impl::get_autograd_meta(var));
+      auto grad_fn = impl::grad_fn_unsafe(var);
+      TORCH_CHECK(diff_view_meta->allow_rebase_history,
+          "The ", diff_view_meta->output_nr_, "th output of ", grad_fn ? grad_fn->name() : "UnknownFunction",
```
If a variable is not a leaf, and it requires grad, grad fn should never be a nullptr.

albanD replied:
Yes, artifact of an older implementation. I'll update it.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -769,47 +771,26 @@ def wrap_output(call):
             # See NOTE [ Autograd View Variables ] in variable.h for details.
```
Should this NOTE be updated from this PR?

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -769,47 +771,26 @@ def wrap_output(call):
-                            view_info_dict[i] = base_name
+
+            if not isinstance(view_info, str):
+                raise TypeError("The view info should be a string for {}".format(base_name))
```
nit: printing `view_info` here would help people debug easier if they make a mistake

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -769,47 +771,26 @@ def wrap_output(call):
+            elif len(differentiable_output_vars) == 1:
+                # Single differentiable output
+                extra = []
+                is_differentiable_view = 'true'
```
nit: The variable assignment here and then interpolation seems slightly pointless. Is it for documentary purposes? Then, even better: can help document the codegen'ed code too by annotating the generated code like `as_view(/* view_info */ {}, /* call */ {}, /* is_differentiable_view */ {})`.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -769,47 +771,26 @@ def wrap_output(call):
+                is_differentiable_view = 'true'
+                wrapped_call = 'as_view({}, {}, {})'.format(view_info, call, is_differentiable_view)
+                return_info = returns[0]
+                if return_info['type'] != 'Tensor':
```
nit: A more defensive way to code this test is to explicitly test for `Tensor` and `TensorList`, and raise an assert failure if the return info matches neither of these. If someone adds support for a new kind of differentiable output (or introduces a bug where a non-Tensor type is marked differentiable), they will then get an error at codegen time, as opposed to a missing overload error at compile time.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -769,47 +771,26 @@ def wrap_output(call):
-                if output_idx in view_info_dict:
-                    stmt = wrap_view_single(output_var, view_info_dict[output_idx])
-                extra_wrapping_stmts.append(stmt)
-            return call, extra_wrapping_stmts
```
This is a very nice simplification, good work.

ezyang reviewed (changes requested):
The nits are just that, nits, don't kill yourself over them. The most substantive concern I have is the lack of modification to the `NOTE` for discussing this situation. In particular, it seems important to me that we explicitly point out that `unbind` is the function that is triggering this situation.

albanD commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -118,6 +127,16 @@ inline std::vector<Tensor> as_view(const Tensor & base, std::vector<Tensor> tens
+  impl::disable_rebase_history(tensor);
+}
+
+inline void disable_rebase_history(std::vector<Tensor> tensors) {
```
I followed what is done for `as_view()` below [here](https://github.com/pytorch/pytorch/blob/968ac49eba8a45b203547a45c38377e05d5b0d24/torch/csrc/autograd/VariableTypeUtils.h#L118). Should I change both to use `ArrayRef<Tensor>` ?

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).


[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).


[ghstack-poisoned]

albanD commented:
Discussion to fix the xla build is in the previous PR in the stack.

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).

Differential Revision: [D19409649](https://our.internmc.facebook.com/intern/diff/D19409649)

[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).

Differential Revision: [D19409649](https://our.internmc.facebook.com/intern/diff/D19409649)

[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).

Differential Revision: [D19409649](https://our.internmc.facebook.com/intern/diff/D19409649)

[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).

Differential Revision: [D19409649](https://our.internmc.facebook.com/intern/diff/D19409649)

[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output.

The codegen change can be found [here](https://github.com/pytorch/pytorch/commit/4bf180caa0e655111b036c8b0419da89fc75671b).

Differential Revision: [D19409649](https://our.internmc.facebook.com/intern/diff/D19409649)

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -295,6 +295,13 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
+///     autograd engine will make an equivalent graph (corresponding to the
+///     view operations) without using the user-provided grad_fn if any output
+///     is modified by an inplace operation. If the provided grad_fn does more
+///     than the backward of the view, then the DifferentiableViewMeta must be
```
In what situations will the grad fn "do more than the backward of the view"? That's only custom autograd right?

albanD replied:
At the moment yes.

ezyang commented on tools/autograd/gen_autograd.py:
```diff
@@ -68,6 +66,21 @@
 RETURNS_VIEWS_OF_INPUT = set(VIEW_FUNCTIONS.keys()).union({'chunk', 'narrow'})
 
+# note: some VIEW_FUNCTIONS are returning multiple views. This means that the engine
+# has to forbid inplace operations on their reults. If such function is only doing views
```
nit: results

ezyang replied:
This comment says "some view functions return multiple views; therefore, we must forbid inplace operations on results." But WHY do I have to forbid inplace operations on the results? The comment doesn't say. Adding to the reader's confusion, you then contradict this statement immediately below by saying we can create each output as output of a separate node. And the second condition "modification of the output will change the memory usage of the backward" isn't actionable: this is more of a gotcha for users to worry about, not something that would affect whether or not a developer would put the backward in this list.

albanD replied:
Updated to make it clearer.

ezyang commented on tools/autograd/gen_autograd.py:
```diff
@@ -68,6 +66,21 @@
+#   backward as the multiple Nodes we create will force the gradient buffer to be created
+#   for each Node, instead of once in the original multi-output Node.
+# A function should NOT be added to this list if:
+# - Your backward function does more than just the backward of the views.
```
Is there ever a situation where functions in core would not faithfully represent the backwards?

albanD replied:
The forward could have some side-effects, not tracked by the autograd, that require a specific backward.

For example a function that promises change in the layout? 
Also, I can see some internal users writing a codegened version of the [Broadcast](https://github.com/pytorch/pytorch/blob/8fd3eaed257884361828217f3aeaabc0296ae906/torch/nn/parallel/_functions.py#L9-L32) we have. It returns a mix of views and not views and does a specific backward because it moves some gradients between GPUs.

ezyang replied:
OK, so put that down in the comment :)

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output. Introduce a whitelist of functions for which we allow this.

Differential Revision: [D19409649](https://our.internmc.facebook.com/intern/diff/D19409649)

[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output. Introduce a whitelist of functions for which we allow this.

Differential Revision: [D19409649](https://our.internmc.facebook.com/intern/diff/D19409649)

[ghstack-poisoned]

albanD committed:
Update on "Make autogen functions correct for multiple outputs and views"


This PR does three things:
- Add a new `allow_rebase_history` flag to the differentiable views. If set, trying to rebase their history will raise an error.
- Make sure that the codegen functions verify this flag before doing inplace operations so that they fail before doing the inplace modification.
- Make sure the codegen functions set this flag properly when we don't support rebasing the history of the output. Introduce a whitelist of functions for which we allow this.

Differential Revision: [D19409649](https://our.internmc.facebook.com/intern/diff/D19409649)

[ghstack-poisoned]
